### PR TITLE
Distributed Automatic SParsity with Fleet

### PR DESCRIPTION
**Pull request by mingxu1067**
### PR types
 New features

### PR changes

APIs
### Describe

1.  Implemented ASP MetaOptimizer in Fleet to enable distributed ASP training.
2.  ASP MetaOptimizer could be the top of other MetaOptimizers.
3.  Added a new attribute, asp, into DistributedStrategy .

Code example:
```python
from paddle.fluid.contrib import sparsity

fleet.init(is_collective=True)
with fluid.program_guard(main_program, startup_program):
    input_data, loss = network()

    optimizer = fluid.optimizer.SGD(learning_rate=0.1)
	strategy = fleet.DistributedStragety()
    strategy.amp = True
    strategy.asp = True
    optimizer = fleet.distributed_optimizer(optimizer, strategy=strategy)
    optimizer.minimize(optimizer, loss, main_program, startup_program)

exe = fluid.Executor(place)
exe.run(startup_program)

# Must call `exe.run(startup_program)` first before calling `sparsity.prune_model`
sparsity.prune_model(place, main_program)

# Start training with ASP workflow.
for epoch in range(total_epochs):
    exe.run(main_program)
```

### CODE REVIEW

**Comment by qingqing01** [comment]
@mingxu1067 Do you have plan to support dygraph and inference?

**Comment by mingxu1067** [comment]
> @mingxu1067 Do you have plan to support dygraph and inference?

We do not have determined plan on DyGraph supporting so far.

File: python/paddle/fluid/contrib/sparsity/asp.py [hunk 1]
```diff
@@ -134,21 +130,36 @@ def prune_model(place,
 
             place = fluid.CUDAPlace(0)
 
+
+
```
Comment (Xreki) [inline python/paddle/fluid/contrib/sparsity/asp.py:141]: 这里空行加的有点多了。

File: python/paddle/fluid/contrib/sparsity/asp.py [hunk 2]
```diff
@@ -134,21 +130,36 @@ def prune_model(place,
...
-                optimizer = decorate(fluid.optimizer.SGD(learning_rate=0.1))
+                # Setup exluded layers out from ASP workflow.
+                # Please note, excluded_layers must be set before calling `optimizer.minimize()`.
+                sparsity.set_excluded_layers(main_program, ["need_dense"])
+
+                optimizer = fluid.optimizer.SGD(learning_rate=0.1)
+                optimizer = fluid.contrib.mixed_precision.decorator.decorate(optimizer )
+                # Calling sparsity.decorate() to wrap minimize() in optimizer, which 
+                # will insert necessary masking operations for ASP workflow.
+                optimizer = sparsity.decorate(optimizer )
```
Comment (Xreki) [inline python/paddle/fluid/contrib/sparsity/asp.py]: 现在是amp和asp一起用有问题？那这个示例里面先不要加amp吧？

Comment (mingxu1067) [reply python/paddle/fluid/contrib/sparsity/asp.py]: ASP與AMP一起使用沒有問題，有成功使用PaddleNLP/bert的AMP + ASP做過sparse finetuning

File: python/paddle/fluid/contrib/sparsity/asp.py (hunk 1 above)
Comment (mingxu1067) [reply python/paddle/fluid/contrib/sparsity/asp.py:141]: Ok, 分散式那邊有沒有甚麼comment? 我一起修改後再commit一次

File: python/paddle/distributed/fleet/meta_optimizers/asp_optimizer.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    def _can_apply(self):
+        if not self.role_maker._is_collective:
+            return False
+
+        if self.user_defined_strategy.asp:
+            return True
+
+        return False
+
+    def _disable_strategy(self, dist_strategy):
```
Comment (sandyhouse) [inline python/paddle/distributed/fleet/meta_optimizers/asp_optimizer.py:47]: set dist_strategy.asp = False ？

File: python/paddle/distributed/fleet/meta_optimizers/asp_optimizer.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        return
+
+    def _enable_strategy(self, dist_strategy, context):
+        return
```
Comment (sandyhouse) [inline python/paddle/distributed/fleet/meta_optimizers/asp_optimizer.py]: set dist_strategy.asp = True ？

**Review by sandyhouse — APPROVED** [review]
LGTM for Fleet strategy

**Review by Xreki — APPROVED** [review]

**Pull request merged**